src/lang/frontend_ast_lower.rs

```diff
@@ -0,0 +1,798 @@
+            });
+        };
+
+        // Check args count
+        if stmt.args.len() > self.params.max_statement_args {
+            return Err(LoweringError::TooManyStatementArgs {
+                count: stmt.args.len(),
+                max: self.params.max_statement_args,
+            });
+        }
+
+        // Convert AST args to BuilderArgs
+        let mut builder = StatementTmplBuilder::new(predicate);
+        for arg in &stmt.args {
+            let builder_arg = self.lower_statement_arg_to_builder(arg)?;
+            builder = builder.arg(builder_arg);
+        }
+
+        // Return builder without calling .desugar() - that will happen later
+        Ok(builder)
+    }
+
+    fn lower_statement_arg_to_builder(
+        &self,
+        arg: &StatementArg,
+    ) -> Result<BuilderArg, LoweringError> {
+        match arg {
+            StatementArg::Literal(lit) => {
+                let value = self.lower_literal(lit)?;
+                Ok(BuilderArg::Literal(value))
+            }
+            StatementArg::Identifier(id) => {
+                // For builder, we just need the wildcard name
+                Ok(BuilderArg::WildcardLiteral(id.name.clone()))
+            }
+            StatementArg::AnchoredKey(ak) => {
+                let key_str = match &ak.key {
+                    AnchoredKeyPath::Bracket(s) => s.value.clone(),
+                    AnchoredKeyPath::Dot(id) => id.name.clone(),
+                };
+                Ok(BuilderArg::Key(ak.root.name.clone(), key_str))
+            }
+        }
+    }
+
+    fn lower_literal(&self, lit: &LiteralValue) -> Result<middleware::Value, LoweringError> {
+        let value = match lit {
+            LiteralValue::Int(i) => middleware::Value::from(i.value),
+            LiteralValue::Bool(b) => middleware::Value::from(b.value),
+            LiteralValue::String(s) => middleware::Value::from(s.value.clone()),
+            LiteralValue::Raw(r) => {
+                // Parse hex string to RawValue (little-endian)
+                let hex_str = &r.hash.value[2..]; // Skip "0x"
+                let raw_value = Self::parse_hex_str_to_raw_value(hex_str)?;
+                middleware::Value::from(raw_value)
+            }
+            LiteralValue::PublicKey(pk) => {
+                // Parse base58 to Point
+                let point: Point = pk
+                    .base58
+                    .parse()
+                    .map_err(|_| LoweringError::InvalidArgumentType)?;
+                middleware::Value::from(point)
+            }
+            LiteralValue::SecretKey(sk) => {
+                // Parse base64 to SecretKey
+                let bytes = deserialize_bytes(&sk.base64)
+                    .map_err(|_| LoweringError::InvalidArgumentType)?;
+                let secret_key = SecretKey::from_bytes(&bytes)
+                    .map_err(|_| LoweringError::InvalidArgumentType)?;
+                middleware::Value::from(secret_key)
+            }
```

Comment:
I would like to propose moving the parsing of these three types to `frontend_ast` so that they are parsed as early as possible.
This will mean that the `frontend_ast::parse::parse_document` would need to return a `Result`.

The reason for this proposal is:
- we can parse these values without knowing the params
- we can parse them in a single pass during the pest pairs tree traversal
- we can return from an error earlier
- the lowering is more complex than the pest tree traversal & parsing, so removing a bit of complexity from the lowerer is good

Reply:
It would also be nice to move the parsing of literal containers to `parse_document` but they depend on params (#436 would fix that)

Reply:
Done, except for the containers.